text/0097-log-data-model.md

```diff
@@ -0,0 +1,968 @@
+
+- Fields stored in the key/value pair lists, which can contain arbitrary values
+  of different types. The keys and values for well-known fields follow semantic
+  conventions for key names and possible values that allow all parties that work
+  with the field to have the same interpretation of the data. See references to
+  semantic conventions for `resource` and `attributes` fields and examples in
+  [Appendix A](#appendix-a-example-mappings).
+
+The reasons for having these 2 kinds of fields are:
+
+- Ability to efficiently represent named top-level fields, which are almost
+  always present (e.g. when using encodings like Protocol Buffers where fields
+  are enumerated but not named on the wire).
+  
+- Ability to enforce types of named fields, which is very useful for compiled
+  languages with type checks.
+
+- Flexibility to represent less frequent data via key/value pair lists. This
+  includes well-known data that has standardized semantics as well as arbitrary
+  custom data that the application may want to include in the logs.
+
+When designing this data model I followed the following reasoning to make a
+decision about when to use use a top-level named field:
+
+- The field needs to be either mandatory for all records or be frequently
+  present in well-known log and event formats (such as `timestamp`) or is
+  expected to be often present in log records in upcoming logging systems (such
+  as `trace_id`).
+
+- The field’s semantics must be the same for all known log and event formats and
+  can be mapped directly and unambiguously to this data model.
+  
+Both of the above conditions were required to give the field a place in the
+top-level structure of the record.
+
+## Log and Event Record Definition
+
+Note: below we use type `any`, which can be a scalar value (number, string or
+boolean), or an array or map of values. Arbitrary deep nesting of values for
+arrays and maps is allowed (essentially allow to represent an equivalent of a
+JSON object).
+
+[Appendix A](#appendix-a-example-mappings) contains many examples that show how
+existing log formats map to the fields defined below. If there are questions
+about the meaning of the field reviewing the examples may be helpful.
+
+Here is the list of fields in a log record:
+
+Field Name     |
+---------------|
+timestamp      |
+trace_id       |
+span_id        |
+trace_flags    |
+severity_text  |
+severity_number|
+short_name     |
+body           |
+resource       |
+attributes     |
+
+Below is the detailed description of each field.
+
+### Field: `timestamp`
+
+Type: Timestamp, uint64 nanosecods since Unix epoch.
+
+Description: Time when the event occurred measured by the origin clock. This
+field is optional, it may be missing the timestamp is unknown.
+
+### Trace Context Fields
+
+#### Field: `trace_id`
+
+Type: byte sequence.
+
+Description: Optional request trace id. Can be set for logs that are part of
+request processing and have an assigned trace id.
```

Comment:
Should we refer to other specs here, where this type is defined? https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#spancontext

Should we also consider using the same casing when describing the field? The spec calls it `TraceId`.

Should this state that it's a 16-byte sequence? That's part of the spec.

Reply:
Good point regarding the casing. Will fix.

I am unable to find a good direct link to TraceId/SpanId definition. Keeping as is for now.

Reply:
https://www.w3.org/TR/trace-context/#trace-id

Reply:
@bogdandrutu Does OpenTelemetry limit its trace ids to W3C trace ids only?

Reply:
Pardon me for intervening. The spec I linked above establishes this:

* `Span`s encapsulate an immutable `SpanContext` that uniquely identifies the `Span`
* The OpenTelemetry `SpanContext` representation conforms to the w3c TraceContext specification. It contains two identifiers - a `TraceId` and a `SpanId` ...
* `TraceId` is a 16-byte array with at least one non-zero byte

It seems to me that the answer to your question is yes, but I'm definitely not the authoritative answer here.

Reply:
Yes we follow the w3c to define IDs. We do provide ways to propagate to/from other formats but the canonical way is w3c

Reply:
After looking at the spec quickly casing was something I wanted to bring up as well.

Preferably just lowercase with no seprators, so `TraceId` becomes `traceid`.

Reply:
@tsloughter the spec currently uses `TraceId` casing. I have no strong preference and I am happy to change here if we also change everywhere else in the spec. I want to make sure we keep it consistent.

Reply:
Yea, I was referring to changing the spec, I just saw casing was mentioned in this thread so thought I'd simply add my support for lowercasing like w3c :)

Reply:
Added link to W3C trace id.